### PR TITLE
[DoctrineBridge] Add decimal form type

### PR DESCRIPTION
**Pull request by fancyweb**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Currently, when an entity field is mapped as a `decimal`, the guessed form type for this field is the `NumberType`. I guess it's also the one that comes naturally to everyone's mind.

The `NumberType` uses the view transformer `NumberToLocalizedStringTransformer` which always transform the submitted string value to an int or a float.

As stated in the Doctrine documentation for the `decimal` type : 
> Values retrieved from the database are always converted to PHP’s string

So when you use the `NumberType`, the string value of your `decimal` field is replaced by its float value. The big issue is that this change of type if detected when Doctrine computes the change sets (strict comparison is done ofc). That mean a lot of unnecessary updates are made.

The new form type I introduce simply extends the `NumberType` but it transform the value back to a string (the model value).

### CODE REVIEW

**Comment by fancyweb** [comment]
I have found a new case that was not handled correctly by my first implementation and that was not detected in the tests because it depends on the database platform.

Fields mapped as "decimal" will always be returned as string. But on some platform the values are rounded when the number of decimals is inferior to the "scale" option.

Eg with inserted float value 8.000 (scale 3) :
- On SQLite => fetched as "8"
- On PostgreSQL => fetched as "8.000"

My solution is to add an option to the new form type to correctly format the number in the model transformer.

- [ ] TODO : automatically set the options in the `DoctrineOrmTypeGuesser`

**Comment by chalasr** [comment]
@xabbuh @stof @webmozart maybe you can help moving forward here?

File: src/Symfony/Bridge/Doctrine/Form/DataTransformer/NumberToStringTransformer.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    private $forceFullScale;
+
+    /**
+     * @var int|null
+     */
+    private $scale;
+
+    /**
+     * @param bool $forceFullScale
```
Comment (Ocramius) [inline src/Symfony/Bridge/Doctrine/Form/DataTransformer/NumberToStringTransformer.php]: Needs documentation

File: src/Symfony/Bridge/Doctrine/Form/DataTransformer/NumberToStringTransformer.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+     * @return string|null
+     */
+    public function transform($value)
+    {
+        if (null === $value) {
+            return null;
+        }
+
+        if (!is_string($value)) {
+            throw new TransformationFailedException('Expected a string.');
```
Comment (Ocramius) [inline src/Symfony/Bridge/Doctrine/Form/DataTransformer/NumberToStringTransformer.php]: Should probably be reflected in the `@param`

File: src/Symfony/Bridge/Doctrine/Form/DataTransformer/NumberToStringTransformer.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                $value = floatval($value);
+            }
+
+            return number_format($value, $this->scale, '.', '');
+        }
+
+        try {
+            return (string) $value;
+        } catch (\Exception $e) {
+            throw new TransformationFailedException();
```
Comment (Ocramius) [inline src/Symfony/Bridge/Doctrine/Form/DataTransformer/NumberToStringTransformer.php]: Previous exception should be passed on here

File: src/Symfony/Bridge/Doctrine/Form/DataTransformer/NumberToStringTransformer.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            if ($valueIsInt) {
+                $value = floatval($value);
+            }
+
+            return number_format($value, $this->scale, '.', '');
+        }
+
+        try {
+            return (string) $value;
+        } catch (\Exception $e) {
```
Comment (Ocramius) [inline src/Symfony/Bridge/Doctrine/Form/DataTransformer/NumberToStringTransformer.php]: Should be `\Throwable` IMO

File: src/Symfony/Bridge/Doctrine/Form/Type/DecimalType.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+use Symfony\Component\Form\Extension\Core\Type\NumberType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class DecimalType extends AbstractType
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
```
Comment (Ocramius) [inline src/Symfony/Bridge/Doctrine/Form/Type/DecimalType.php]: `force_full_scale` option to be documented above

Comment (Ocramius) [reply src/Symfony/Bridge/Doctrine/Form/Type/DecimalType.php]: `scale` too

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by Ocramius** [comment]
About https://github.com/symfony/symfony/pull/24793#issuecomment-372703184, I just discussed it with @webmozart and it is not fixable in `doctrine/dbal` without introducing a BC break in [`DecimalType`](https://github.com/doctrine/dbal/blob/799085fac21c9759c1fcef828586418723ad063f/lib/Doctrine/DBAL/Types/DecimalType.php#L28-L34).

There is a solution, which is making the `AbstractPlatform` configurable, but that's a whole lot of additions for a very edge-case scenario.

If the whole issue drills down to avoiding superfluous updates at DB level, I don't think that problem should be tackled here, and instead a change in `doctrine/dbal` is to be proposed.

**Comment by webmozart** [comment]
Working on finishing this as part of EUFOSSA.

**Comment by fancyweb** [comment]
I agree that Doctrine should return coherent values between platforms for this case.

I think we can just use my initial commit with the new Form type, it will still be better than now and will be perfect once Doctrine is ready.

@webmozart Do you mean you are gonna work on a fix on Doctrine side?

**Comment by fancyweb** [comment]
Closing for https://github.com/symfony/symfony/pull/30893.

**Comment by webmozart** [comment]
@webmozart No, I provided some thoughts on how to implement this in #30893 though. :)

**Comment by fancyweb** [comment]
Was an issue opened on Doctrine side to track this problem?

**Comment by webmozart** [comment]
@fancyweb Not yet, I'd appreciate if you could do so!

**Pull request closed without merging**